From ca3dac299f1139a83ce29f37f800a5c94e0b9a67 Mon Sep 17 00:00:00 2001
From: Richard Zhu <hongxing.zhu@nxp.com>
Date: Wed, 25 Oct 2017 10:48:48 +0800
Subject: [PATCH 2688/5242] MLK-16684-3 ata: imx: enable imx8qm sata

commit  9ffab491b4e44e00b706d646fafceb5c2d8ffaee from
https://source.codeaurora.org/external/imx/linux-imx.git

enable sata on imx8qm.
sata function is relied on the usage of pcie ports.

BuildInfo:
- SCFW 9559d5ec, IMX-MKIMAGE 06bc2767, ATF
- U-Boot 2017.03-imx_v2017.03_4.9.51_imx8_beta1+g6dc7b0f

Reviewed-by: Frank Li <frank.li@nxp.com>
Signed-off-by: Richard Zhu <hongxing.zhu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/ata/Kconfig    |    2 +-
 drivers/ata/ahci_imx.c |    5 ++++-
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/ata/Kconfig b/drivers/ata/Kconfig
index 39b181d6..d284d81 100644
--- a/drivers/ata/Kconfig
+++ b/drivers/ata/Kconfig
@@ -156,7 +156,7 @@ config AHCI_ST
 
 config AHCI_IMX
 	tristate "Freescale i.MX AHCI SATA support"
-	depends on MFD_SYSCON && (ARCH_MXC || COMPILE_TEST)
+	depends on MFD_SYSCON && (ARCH_MXC || ARCH_FSL_IMX8QM || COMPILE_TEST)
 	depends on (HWMON && (THERMAL || !THERMAL_OF)) || !HWMON
 	help
 	  This option enables support for the Freescale i.MX SoC's
diff --git a/drivers/ata/ahci_imx.c b/drivers/ata/ahci_imx.c
index 393ab35..d99fe06 100644
--- a/drivers/ata/ahci_imx.c
+++ b/drivers/ata/ahci_imx.c
@@ -22,6 +22,8 @@
 #include <linux/platform_device.h>
 #include <linux/regmap.h>
 #include <linux/ahci_platform.h>
+#include <linux/gpio.h>
+#include <linux/of_gpio.h>
 #include <linux/of_device.h>
 #include <linux/of_gpio.h>
 #include <linux/mfd/syscon.h>
@@ -780,7 +782,8 @@ static void ahci_imx_error_handler(struct ata_port *ap)
 
 	ahci_error_handler(ap);
 
-	if (!(imxpriv->first_time) || ahci_imx_hotplug)
+	if (!(imxpriv->first_time) || ahci_imx_hotplug
+			|| (imxpriv->type == AHCI_IMX8QM))
 		return;
 
 	imxpriv->first_time = false;
-- 
1.7.9.5

